### PR TITLE
Tactic Assignment Oscillations

### PR DESCRIPTION
**Pull request by itsarune**
### Description

STP now prefers old tactic assignments as long as the total cost difference between the old and new is within a threshold. The threshold is the highest possible that continues to make the tests pass.

- ShootOrPassPlayTest occasionally works

### Testing Done

Quantified the tactic swaps for ShootOrPassPlay before and after these changes. There two videos compare ShootOrPassPlay before/after the changes and the threshold system seems to make it more stable.

new-> https://drive.google.com/file/d/1Wkpb2k16an8Sgp0cz0d4hlhsuK8aw-z1/view?usp=sharing
old-> https://drive.google.com/file/d/1WzNDvT_wpuDycu0BwNspJ5XwoxqWtRDY/view?usp=sharing

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/hl/stp/stp.cpp [hunk 1]
```diff
@@ -345,5 +369,13 @@ std::map<std::shared_ptr<const Tactic>, Robot> STP::assignRobotsToTactics(
         robots = remaining_robots;
     }
 
+    if (recalculate_tactic_assignment || (old_assignment > (new_assignment + NEW_TACTIC_ASSIGNMENT_THRESHOLD)))
+    {
+        robot_tactic_assignment = *new_tactic_assignment;
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/stp.cpp]: I think raw pointers need to be avoided. I added a `delete new_tactic_assignment` here and it changed things (for the better maybe)

File: src/software/ai/hl/stp/stp.cpp [hunk 2]
```diff
@@ -219,19 +219,52 @@ std::map<std::shared_ptr<const Tactic>, Robot> STP::assignRobotsToTactics(
...
+
     if (goalie_robot && automatically_assign_goalie)
     {
-        robot_tactic_assignment.emplace(goalie_tactic, goalie_robot.value());
+        ++new_tactics;
+        if (robot_tactic_assignment.find(goalie_tactic) == robot_tactic_assignment.end())
+        {
+            recalculate_tactic_assignment = true;
+        }
+        new_tactic_assignment.emplace(goalie_tactic, goalie_robot.value());
```
Comment (itsarune) [inline src/software/ai/hl/stp/stp.cpp]: Looking for suggestions to make this code clearer.
Basically, force re-calculate the tactic assignment for this tick if:
- requested tactic vector size is different from previous tactic vector size
- requested tactic wasn't assigned previously

File: src/software/ai/hl/stp/stp.cpp (hunk 1 above)
Comment (itsarune) [reply src/software/ai/hl/stp/stp.cpp]: no more raw pointers

**Comment by akhilveeraghanta** [comment]
@itsarune whats the latest on this? do we want to close this PR? or is this blocked on something else?

**Comment by itsarune** [comment]
I did forget about this, but the global path planner will probably re-structure some of this code, so probably just makes sense closing this

**Pull request closed without merging**